distribution/metadata/v2_metadata_service.go

```diff
@@ -15,10 +18,24 @@ type V2MetadataService struct {
 
 // V2Metadata contains the digest and source repository information for a layer.
 type V2Metadata struct {
+	AuthConfigHash   string
 	Digest           digest.Digest
 	SourceRepository string
 }
 
+// HashAuthConfig returns an sha256 sum of the "authConfig" if successful.
+func HashAuthConfig(authConfig *types.AuthConfig) (string, error) {
+	if authConfig == nil {
+		return "", nil
+	}
+	buf, err := json.Marshal(authConfig)
```

Comment:
I think something along the following lines would also work:

```go
hashInput := authConfig.Username
if authConfig.IdentityToken != "" {
        hashInput = authConfig.IdentityToken
} else if authConfig.RegistryToken != "" {
        hashInput = authConfig.RegistryToken
}
return digest.FromBytes([]byte(hashInput)).String(), nil
```

It's still important to do the hashing, because we don't want to put bare tokens in the distribution metadata store.